### PR TITLE
OCPBUGS-32183: catalog-operator: delete catalog pods stuck in Terminating state due to unreachable node

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**

**Motivation for the change:**

**Architectural changes:**

**Testing remarks:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Bug fixes are accompanied by regression test(s)
- [ ] e2e tests and flake fixes are accompanied evidence of flake testing, e.g. executing the test 100(0) times
- [ ] tech debt/todo is accompanied by issue link(s) in comments in the surrounding code
- [ ] Tests are comprehensible, e.g. Ginkgo DSL is being used appropriately
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive
- [ ] Tests marked as `[FLAKE]` are truly flaky and have an issue
- [ ] Code is properly formatted

### CODE REVIEW

File: pkg/controller/registry/reconciler/grpc.go [hunk 1]
```diff
@@ -338,13 +339,30 @@ func isRegistryServiceStatusValid(source *grpcCatalogSourceDecorator) (bool, err
 }
 
 func (c *GrpcRegistryReconciler) ensurePod(logger *logrus.Entry, source grpcCatalogSourceDecorator, serviceAccount *corev1.ServiceAccount, overwrite bool) error {
-	// currentLivePods refers to the currently live instances of the catalog source
-	currentLivePods := c.currentPods(logger, source)
-	if len(currentLivePods) > 0 {
+	// currentPods refers to the current pod instances of the catalog source
+	currentPods := c.currentPods(logger, source)
+
+	tmpPods := currentPods[:0]
```
Comment (perdasilva) [inline pkg/controller/registry/reconciler/grpc.go:345]: nifty! didn't know about this one

Comment (joelanford) [reply pkg/controller/registry/reconciler/grpc.go:345]: Actually, this reminds me there's a new kid on the block (`slices.DeleteFunc`). I'll switch to that as the intent is more clear.

**Comment by grokspawn** [comment]
would love to better understand the need for the errors --> pkgerrors change, to know if that's a pattern we should look for elsewhere.

**Comment by joelanford** [comment]
`errors` is the name of a very popular package in the Go standard library.

This package was authored prior to standard library Go having `fmt.Errorf("blah blah: %w", err)`. Before Go supported that, `pkg/errors.Wrapf(err, "message")` was a popular way to wrap errors.

I made this change to give the `errors` name to stdlib `errors` over the third-party errors package.

File: pkg/controller/registry/reconciler/grpc.go [hunk 2]
```diff
@@ -523,6 +545,29 @@ func imageChanged(logger *logrus.Entry, updatePod *corev1.Pod, servingPods []*co
 	return false
 }
 
+func isPodDead(pod *corev1.Pod) bool {
+	for _, check := range []func(*corev1.Pod) bool{
```
Comment (grokspawn) [inline pkg/controller/registry/reconciler/grpc.go:549]: This is a clever way to provide an extensible approach to a series of checks... but do we need it?  It seems we would have the same benefits by doing `s/isPodDead/IsPodDeletedByTaintManager/g`

Comment (joelanford) [reply pkg/controller/registry/reconciler/grpc.go:549]: I'm somewhat confident we'll find another way for pods to be dead. We've seen similar issues in `operator-lib` leader for life.

So I figured I'd make things super easy for ourselves next time around.

Comment (tmshort) [reply pkg/controller/registry/reconciler/grpc.go:549]: Yeah, do we think we'll need to expand this in the future?

Comment (grokspawn) [reply pkg/controller/registry/reconciler/grpc.go:549]: Well, I ain't agin' it, but I generally like to supply it when it's needed. 
/lgtm
from me

Comment (tmshort) [reply pkg/controller/registry/reconciler/grpc.go:549]: I'm not objecting... just asking.

**Pull request merged**